### DISCUSSION TITLE
Additional response model does not validate response returned by endpoint callable

### DISCUSSION THREAD

**Discussion by ondrejkajinek** [fastapi/fastapi#6446] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi.responses import JSONResponse
from pydantic import BaseModel
import uvicorn


class Item(BaseModel):
    id: str
    value: str


class Message(BaseModel):
    message: str


app = FastAPI()


@app.get("/items/{item_id}", response_model=Item, responses={404: {"model": Message}})
async def read_item(item_id: str):
    if item_id == "foo":
        return {"id": "foo", "value": "there goes my hero"}
    else:
        return JSONResponse(status_code=404, content={"MESSAGE": "Item not found"})


if __name__ == "__main__":
    uvicorn.run("__main__:app", host="0.0.0.0", port=8080)
```

### Description

* curl localhost:8080/items/fooo
* It returns `{"MESSAGE":"Item not found"}`
* 500 should be returned as response content is not valid `Message`.

### Operating System

Linux

### FastAPI Version

0.70.0

### Python Version

3.9.7

**Comment by NinaHwang** [comment, marked as answer]
Because in this case, you are directly returning JSONResponse. Here, "responses={404: {"model": Message}}" is for creating response in OpenAPI, not validating the response. It just shows how it supposed to look.

**Comment by ondrejkajinek** [comment]
Ok, is there a way to define response model for a specific status code? Here is my use-case (really simple one): when some entity is in db, you return it in some specific structure => `response_model` passed to api route. When the entity is not found, you return 404 with some error message that should also have some model because I want to have unified error messages accross all endpoints.

something like this (imports and definition of `Item` and `Message` omitted):

```
@app.get(
    "/items/{item_id}",
    response_model=Item,
    responses={
        404: {
            "response_model": Message,
        },
    },
)
async def read_item(item_id: str, response: Response):
    if item_id == "foo":
        return {"id": "foo", "value": "there goes my hero"}
    else:
        # this would be validated with `Message` model, so {"Message": "Itme not found"} would be invalid
        response.status_code = 404
        return {"message": "Item not found"}

```

Thanks in advance.

**Comment by NinaHwang** [comment]
In that case, I'd store the exception into a variable and use it across all endpoints.
E.g.
```python
from fastapi import HTTPException

NOT_FOUND_ERROR = HTTPException(
    status_code=404,
    detail={"message": "Item not found"}
)

class Item(BaseModel):
    id: str
    value: str

@app.get("/items/{item_id}", response_model=Item)
async def read_item(item_id: str):
    if item_id != "foo":
        raise NOT_FOUND_ERROR
    return {"id": "foo", "value": "there goes my hero"}
```
You can also add OpenAPI response, if you need that as well.
```python
class ErrorMessage(BaseModel):
    message: str
class ErrorResponse(BaseModel):
    detail: ErrorMessage

@app.get("/items/{item_id}", response_model=Item, responses={404: {"model": ErrorResponse}})
async def read_item(item_id: str):
    if item_id != "foo":
        raise NOT_FOUND_ERROR
    return {"id": "foo", "value": "there goes my hero"}
```

**Comment by ondrejkajinek** [comment]
Ok, thanks, it's clean now.
